software/spmd/pc_profiler/profiler.S

```diff
@@ -0,0 +1,232 @@
+
+  Memory layout:
+  DRAM --> 16MB (configurable) for maintaing a histogram for N tiles. 
+
+  Registers used:
+  The program being profiled is free to use any registers within if there are no other imposed constraints.
+  The profiler uses the following registers - 
+  t0 --> Base address of the histogram array (actual value stored in dmem[1])
+  t1 --> Scratchpad (actual value stored in dmem[0])
+*/
+
+#include "bsg_manycore_arch.h"
+#include "bsg_manycore_asm.h"
+
+// Useful to control the amount of space required to simulate the use of the profiler.
+// We want 16MB when running on actual hardware.
+#ifndef HIST_SPACE
+#define HIST_SPACE 16777216
+#endif
+
+#ifndef PROFILE
+#define PROFILE 0
+#endif
+
+#if PROFILE > 0
+/* 
+  Data section:
+  Defines a 2 element (8 byte) array to store the base pointer of the histogram array and for use by the interrupt handler to spill some registers
+  dmem[0] = histogram base pointer, temporary storage
+  dmem[1] = temporary storage
+*/
+.section .dmem.interrupt, "aw"
+  _interrupt_arr: .space 8, 0x0
+
+/*
+  DRAM section:
+  Defines a 16MB (array size is configurable) chunk in DRAM to store the histogram.
+
+  Lingering question(s):
+  1. Assigned space per PC (i.e bin capacity) as 32-bits. Is this enough, too much or too less?
+  2. Base pointer in DRAM (for now it is mapped to start of DRAM) [Consider static assignment in linker]
+*/
+.section .dram, "aw"
+.globl _histogram_arr
+  _histogram_arr: .space HIST_SPACE, 0x0
+#endif
+
+/*
+  Startup:
+  This is very similar to the code in software/spmd/common/crt.S and adds only the interrupt service routine (ISR) and sets up the DMEM with the histogram base address. 
+*/
+.section .crtbegin
+.globl _start
+
+#if PROFILE > 0
+/*
+  Remote interrupt handler (should map to address 0)
+*/
+_remote_interrupt:
+  j _remote_interrupt_handler
+
+/*
+  Trace interrupt handler (should map to address 4)
+  Clears the pending trace interrupt bit in the MIP CSR and atomically increments the histogram bin count based on the MEPC CSR value
+
+  The bin index in the histogram can be found using the following equation index = base ptr + EPC
+  (Although EPC is 4-byte aligned, we are assigning 4 bytes per bin, so we don't need to do EPC >> 2)
+
+  Lingering concerns:
+  Since we are using the MEPC, we don't have the PC of the first instruction in the program. 
+  (Probably does not matter because it's just 1 datapoint)
+*/
+_trace_interrupt:
+  // Save a register to dmem[1]
+  sw t0, 4(zero)
+
+  // clear mip.trace
+  li t0, 0x20000
+  csrrc zero, mip, t0
+
+  // load address of histogram array
+  lw t0, 0(zero)
+  // Store another register to dmem[0]
+  sw t1, 0(zero)
+
+  // read the mepc register
+  csrr t1, mepc
+  // find the correct index in the histogram array
+  add t0, t0, t1
+
+  // atomically increment the value at the computed index
+  li t1, 1
+  // We don't care about the result hence assigned to zero register
+  amoadd.w.aq zero, t1, (t0)
```

Comment:
We should have a justification for which combination of .aq and .rl we use. My opinion here is that we don't need either .aq or .rl because we are not using the address as a lock, but as a histogram, and that this is implicitly doing a memory fence after we do the amoadd operation, which is possibly responsible for the slow down you are seeing.

Reply:
That makes a lot of sense. Without the .aq one can just send the amoadd and forget about it since the return register is zero (Need to confirm though if there are no return packets when the rd register is the zero register)

Reply:
You would want the aq if you require the amoadd to complete before any following memory operation in the program. That's not the case here.